### PR TITLE
add stripDomain to template function

### PR DESCRIPTION
**Pull request by fatsheep9146**
This PR add a new function 'stripDomain' to template functions, which is proposed in #9932 .

The function will split string into host and port, then if host is not IPv4 / IPv6, removes all the domain path elements, leaving only the first one.

### CODE REVIEW

**Comment by roidelapluie** [comment]
hello,

I do not think that stripDomain should strip the port. It is easy enough for users to chain them.

**Comment by fatsheep9146** [comment]
> hello,
> 
> I do not think that stripDomain should strip the port. It is easy enough for users to chain them.

ok, so in your opinion, the following expressions should be evaluated like this, 
```
{{ "foo.example.com" | stripDomain }} -> foo
{{ "foo.example.com:12345" | stripDomain }} -> foo:12345
{{ "192.0.2.1:12345" | stripDomain }} -> 192.0.2.1:12345
```
Am I understanding this correctly?

**Comment by roidelapluie** [comment]
Yes

**Comment by fatsheep9146** [comment]
> Yes

Ok, I've already changed the logic.

File: template/template.go [hunk 1]
```diff
@@ -192,6 +192,29 @@ func NewTemplateExpander(
...
+					host = hostPort
+				}
+				ip := net.ParseIP(host)
+				if ip != nil {
+					// if host part is ipv4 or v6, return the hostPort
+					ret = hostPort
+					return
+				}
+				ss := strings.Split(host, ".")
+				if len(ss) == 0 {
```
Comment (roidelapluie) [inline template/template.go:208]: I don't think this can happen. The string will always be the first part.

File: template/template.go [hunk 2]
```diff
@@ -192,6 +192,29 @@ func NewTemplateExpander(
...
+					ret = hostPort
+					return
+				}
+				ret = ss[0]
+				if port != "" {
+					ret = ss[0] + ":" + port
```
Comment (roidelapluie) [inline template/template.go:214]: Let's use net.JoinHostPort to handle ipv6 as well.

File: template/template.go [hunk 3]
```diff
@@ -192,6 +192,29 @@ func NewTemplateExpander(
...
 			},
+			"stripDomain": func(hostPort string) (ret string) {
+				host, port, err := net.SplitHostPort(hostPort)
+				if err != nil {
+					// if hostPort does not contains port, then host equals to hostPort
+					host = hostPort
+				}
+				ip := net.ParseIP(host)
+				if ip != nil {
+					// if host part is ipv4 or v6, return the hostPort
```
Comment (roidelapluie) [inline template/template.go:203]: This comment is not needed.

File: template/template.go [hunk 4]
```diff
@@ -192,6 +192,29 @@ func NewTemplateExpander(
 				}
 				return host
 			},
+			"stripDomain": func(hostPort string) (ret string) {
+				host, port, err := net.SplitHostPort(hostPort)
+				if err != nil {
+					// if hostPort does not contains port, then host equals to hostPort
```
Comment (roidelapluie) [inline template/template.go:198]: This comment is not needed.

Comment (fatsheep9146) [reply template/template.go]: I've already removed it

File: template/template.go (hunk 3 above)
Comment (fatsheep9146) [reply template/template.go]: I've already removed it

File: template/template.go (hunk 2 above)
Comment (fatsheep9146) [reply template/template.go]: I've already used  net.JoinHostPort

File: template/template.go (hunk 1 above)
Comment (fatsheep9146) [reply template/template.go]: I've already removed it

File: template/template.go [hunk 5]
```diff
@@ -192,6 +192,23 @@ func NewTemplateExpander(
...
+				host, port, err := net.SplitHostPort(hostPort)
+				if err != nil {
+					host = hostPort
+				}
+				ip := net.ParseIP(host)
+				if ip != nil {
+					ret = hostPort
+					return
+				}
+				ss := strings.Split(host, ".")
```
Comment (roidelapluie) [inline template/template.go:205]:
```suggestion
				host, _, _ = strings.Cut(host, ".")
				if port != "" {
				   return net.JoinHostPort(host, port)
				}
				return host
```

File: template/template.go [hunk 6]
```diff
@@ -192,6 +192,23 @@ func NewTemplateExpander(
 				}
 				return host
 			},
+			"stripDomain": func(hostPort string) (ret string) {
```
Comment (roidelapluie) [inline template/template.go:195]: Can we avoid the named return?

Comment (fatsheep9146) [reply template/template.go]: Ok. I've already changed this.

File: docs/configuration/template_reference.md [hunk 7]
```diff
@@ -76,6 +76,7 @@ versions.
 | graphLink  | expr | string | Returns path to graph view in the [expression browser](https://prometheus.io/docs/visualization/browser/) for the expression. |
 | tableLink  | expr | string | Returns path to tabular ("Table") view in the [expression browser](https://prometheus.io/docs/visualization/browser/) for the expression. |
 | parseDuration | string | float | Parses a duration string such as "1h" into the number of seconds it represents. |
+| stripDomain | string | string | Split string into host and port, then if host is not IPv4 / IPv6, removes all the domain path elements, leaving only the first one and the port. |
```
Comment (roidelapluie) [inline docs/configuration/template_reference.md:79]:
```suggestion
| stripDomain | string | string | Removes the demain part of a FQDN. Leaves port untouched. |
```

File: template/template.go [hunk 8]
```diff
@@ -192,6 +192,24 @@ func NewTemplateExpander(
 				}
 				return host
 			},
+			"stripDomain": func(hostPort string) string {
+				ret := ""
```
Comment (roidelapluie) [inline template/template.go:196]:
```suggestion
				ret := ""
```

File: template/template.go [hunk 9]
```diff
@@ -192,6 +192,24 @@ func NewTemplateExpander(
...
+				if ip != nil {
+					ret = hostPort
+					return ret
+				}
+				ss := strings.Split(host, ".")
+				ret = ss[0]
+				if port != "" {
+					ret = net.JoinHostPort(ss[0], port)
+				}
+				return ret
```
Comment (roidelapluie) [inline template/template.go:211]:
```suggestion
				host, port, err := net.SplitHostPort(hostPort)
				if err != nil {
					host = hostPort
				}
				ip := net.ParseIP(host)
				if ip != nil {
					return hostPort
				}
				host = strings.Split(host, ".")[0]
				if host != "" {
					return net.JoinHostPort(host, port)
				}
				return host
```

Comment (roidelapluie) [reply template/template.go:211]: Let's get rid of ret.

File: docs/configuration/template_reference.md (hunk 7 above)
Comment (fatsheep9146) [reply docs/configuration/template_reference.md:79]: This summary is much better :). but "demain" or "domain", is it a typo?

File: template/template.go (hunk 9 above)
Comment (fatsheep9146) [reply template/template.go:211]: Ok, I've already added this suggestion to batch

**Comment by fatsheep9146** [comment]
@roidelapluie hi, Is there anything more I should do for this pr？

File: template/template.go [hunk 10]
```diff
@@ -192,6 +192,24 @@ func NewTemplateExpander(
...
 			},
+			"stripDomain": func(hostPort string) string {
+				ret := ""
+				host, port, err := net.SplitHostPort(hostPort)
+				if err != nil {
+					host = hostPort
+				}
+				ip := net.ParseIP(host)
+				if ip != nil {
+					ret = hostPort
```
Comment (roidelapluie) [inline template/template.go:203]:
```suggestion
					return hostPort
```

File: template/template.go [hunk 11]
```diff
@@ -192,6 +192,24 @@ func NewTemplateExpander(
...
+					return ret
+				}
+				ss := strings.Split(host, ".")
+				ret = ss[0]
+				if port != "" {
+					ret = net.JoinHostPort(ss[0], port)
```
Comment (roidelapluie) [inline template/template.go:209]:
```suggestion
					return net.JoinHostPort(host, port)
```

File: template/template.go (hunk 9 above)
Comment (roidelapluie) [inline template/template.go:211]:
```suggestion
				return host
```

**Comment by fatsheep9146** [comment]
@roidelapluie I have rewrote the codes as you suggested. Please help check it again. Thanks

**Comment by fatsheep9146** [comment]
Will you review this pr again？ thanks. @roidelapluie

**Comment by roidelapluie** [comment]
Thanks!

**Comment by baryluk** [comment]
Thanks!

**Pull request merged**